src/kudu/consensus/leader_election.cc

```diff
@@ -1062,60 +1077,59 @@ std::pair<bool, bool> FlexibleVoteCounter::IsDynamicQuorumSatisfied() const {
     VLOG_WITH_PREFIX(1)
-        << "Election term immediately succeeds term of the last known leader."
+        << "Election term immediately succeeds term of the last known leader"
+        << " or continuity in terms is not required."
         << " Election term: " << election_term_
-        << " lkl_term: " << last_known_leader.election_term()
+        << " lkl term: " << last_known_leader.election_term()
         << " lkl uuid: " << last_known_leader.uuid()
         << " lkl region: " << last_known_leader_region
         << " continuity_not_required: " << continuity_not_required;
-    std::pair<bool, bool> result =
-        IsMajoritySatisfiedInRegion(last_known_leader_region);
-    if (FLAGS_srd_strict_leader_election_quorum) {
-      std::pair<bool, bool> majority_result =
-          IsMajoritySatisfiedInMajorityOfRegions();
-      result = std::make_pair<>(
-          result.first && majority_result.first,
-          result.second && majority_result.second);
-    }
-    if (FLAGS_include_candidate_region &&
-        candidate_region != last_known_leader_region) {
-      std::pair<bool, bool> cresult =
-          IsMajoritySatisfiedInRegion(candidate_region);
-      result = std::make_pair<>(
-          result.first && cresult.first,
-          result.second && cresult.second);
+    result =
+        AreMajoritiesSatisfied({last_known_leader_region}, candidate_region);
+  } else {
+    // Step 4: Find possible leader regions at every term greater than last
+    // known leader's term. Computes possible successor regions until next term
+    // is the current election's term or the quorum converges to pessimistic
+    // quorum.
+    //
+    // If we come here, it is guaranteed that none of the current votes that
+    // we have received (including our own) help us figure out last known
+    // LEADER, or we are running an election where there has been a period
+    // of confusion, i.e. we are running an election at a term which does not
+    // immediately follow Last Known Leader. This is not expected to happen
+    // due to PreVote being used all the time, but in some cases ForcedElections
+    // (StartElection) with failure can lead to analysis paralysis term.
+    //
+    // So our hail mary is: If by analyzing the votes, we can find potential
+    // leader regions, we can intersect with those potential regions in the
+    // hope that it will be less than pessimistic quorum.
+    result = ComputeElectionResultFromVotingHistory(
+        last_known_leader, last_known_leader_region, candidate_region);
+
+    auto now = std::chrono::system_clock::now();
+    long time_elapsed_secs =
+        std::chrono::duration_cast<std::chrono::seconds>(now - creation_time_)
+            .count();
+
+    // Case: Pessimistic quorum satisfaction is possible and we still have time
+    // to wait for it
+    if (pessimistic_result.second &&
+        time_elapsed_secs < FLAGS_wait_for_pessimistic_quorum_secs) {
```

Comment:
Overlapping with pessimistic quorum if we've not reached the timeout yet. Default timeout is 10sec which might be too much (we should probably take election timeout into account but that might make things complicated). Looking for suggestions for what to set this timeout to.

Reply:
Will move this before computing from voting history to save cycles